torchvision/prototype/utils/_internal.py

```diff
@@ -126,3 +144,70 @@ def __delattr__(self, item: Any) -> NoReturn:
 
     def __repr__(self) -> str:
         return make_repr(type(self).__name__, self.items())
+
+
+def kwonly_to_pos_or_kw(
+    *,
+    param_map: Optional[Dict[str, str]] = None,
```

Comment:
It's not clear to me what this does, could you detail?

Reply:
As explained in the docstring, this is used to map the parameter names. Let's say you had a function

```python
def foo(bar):
    pass
```

and the new signature looks like 

```python
def foo(*, baz):
   pass
```

If we don't have a map for `"baz": "bar"`, the warning would read something like "The parameter 'baz' is deprecated" although the user only ever used `bar`. I think it is easier to understand for the user this way, but no strong opinion.

Reply:
Do we have such a scenario where we plainly remove a parameter?

I believe we only replace `pretrained` with `weights`. In this case, users should get a deprecation warning telling them "Use weights, not pretrained". Once they address this, I'm not sure the mapping is relevant?

Reply:
The problem is that we do both, convert to keyword only and rename the parameter. For example, removing the map will result in the following warnings:

```python
from torchvision.prototype import models

models.alexnet(True)
```

```
UserWarning: Using 'weights' as positional parameter(s) is deprecated. Please use it/them as keyword parameter(s) instead.
UserWarning: The parameter 'pretrained' is deprecated, please use 'weights' instead.
```

which I think is quite confusing. Without manually telling the decorator that there was a `pretrained` parameter before, it can't generate a fitting warning.

Reply:
Thanks for the explanation @pmeier , I agree the 2 warnings might be confusing

Reply:
I think this mapping of parameters is something also proposed by you Nicolas in our early discussions when I had changed the order of parameters. Could you clarify if your concern is around the way this is implemented or over the need to inform people to stop using positional arguments?

Reply:
Let me see if I can refactor the warning generation so both warnings together are not confusing.

Reply:
Sorry, I meant I agree with @pmeier 's statement 

> which I think is quite confusing

and I thus agree that this `param_map` is useful

Reply:
I have a local commit that emits the following warnings in the four use cases without using the manual map:

- `alexnet(True)`
    ```
    UserWarning: Using 'weights' as positional parameter(s) is deprecated. Please use it/them as keyword parameter(s) instead.
    UserWarning: The current behavior is equivalent to passing `weights=AlexNet_Weights.ImageNet1K_V1`. You can also use `weights='default'` to get the most up-to-date weights.
    ```
- `alexnet(False)`
    ```
    UserWarning: Using 'weights' as positional parameter(s) is deprecated. Please use it/them as keyword parameter(s) instead.
    UserWarning: The current behavior is equivalent to passing `weights=None`. You can also use `weights='default'` to get the most up-to-date weights.
    ```
- `alexnet(pretrained=True)`
    ```
    UserWarning: The parameter 'pretrained' is deprecated, please use 'weights' instead. The current behavior is equivalent to passing `weights=AlexNet_Weights.ImageNet1K_V1`. You can also use `weights='default'` to get the most up-to-date weights.
    ```
- `alexnet(pretrained=False)`
    ```
    UserWarning: The parameter 'pretrained' is deprecated, please use 'weights' instead. The current behavior is equivalent to passing `weights=None`. You can also use `weights='default'` to get the most up-to-date weights.
    ```

Happy with them?